python/ray/llm/_internal/batch/stages/sglang_engine_stage.py

```diff
@@ -0,0 +1,367 @@
+
+        request = SGLangEngineRequest(
+            request_id=self.request_id,
+            idx_in_batch=row[self.idx_in_batch_column],
+            text=text,
+            input_ids=input_ids,
+            sampling_params=sampling_params,
+        )
+        self.request_id += 1
+        return request
+
+    async def generate_async(
+        self, row: Dict[str, Any]
+    ) -> Tuple[SGLangEngineRequest, Dict[str, Any]]:
+        """Process a single request.
+
+        Args:
+            request: The request.
+
+        Returns:
+            A tuple of index in batch, request output and bypassed custom fields.
+        """
+        request = await self._prepare_llm_request(row)
+
+        async with self.semaphore:
+            output = await self._generate_async(request)
+
+        output_data = SGLangOutputData.from_sglang_engine_output(output)
+        return request, output_data.model_dump()
+
+    async def _generate_async(self, request: SGLangEngineRequest) -> Any:
+        """Process a single request.
+
+        Args:
+            request: The request.
+
+        Returns:
+            The output of the request.
+        """
+
+        # Send the request to the LLM engine.
+        output = await self.engine.async_generate(
+            prompt=request.text,
+            sampling_params=request.sampling_params,
+            input_ids=request.input_ids,
+        )
+        output["prompt"] = request.text
+        output["prompt_token_ids"] = request.input_ids
+        return output
+
+    def shutdown(self):
+        """Shutdown the SGLang engine."""
+        if hasattr(self.engine, "shutdown"):
+            logger.info("Shutting down SGLang engine")
+            self.engine.shutdown()
+
+
+class SGLangEngineStageUDF(StatefulStageUDF):
+    def __init__(
+        self,
+        data_column: str,
+        model_path: str,
+        engine_kwargs: Dict[str, Any],
+        task_type: SGLangTaskType = SGLangTaskType.GENERATE,
+        max_pending_requests: Optional[int] = None,
+    ):
+        """
+        Initialize the SGLangEngineStageUDF.
+
+        Args:
+            data_column: The data column name.
+            model_path: The path to the model to use for the SGLang engine.
+            engine_kwargs: The kwargs to pass to the SGLang engine.
+            task_type: The task to use for the SGLang engine (e.g., "generate", "embed", "reward").
+            max_pending_requests: The maximum number of pending requests. If None,
+                it will be set to a default value based on engine settings.
+        """
+        super().__init__(data_column)
+        self.model_path = model_path
+
+        # Setup SGLang engine kwargs.
+        self.task_type = task_type
+        self.engine_kwargs = self.normalize_engine_kwargs(task_type, engine_kwargs)
+
+        # Set up the max pending requests.
+        # Use a reasonable default value if not specified
+        self.max_pending_requests = max_pending_requests or 128
```

Comment:
We should just disable the pending queue if this is not specified.

Reply:
done